File: core/cyp/observability/metrics.py
"""运行指标：结果计数 + 成交/滑点聚合 + SLO（审批时延/滑点分布/下单成功率）。

供 GET /api/metrics 与巡检（M6：SLO 细化）。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from decimal import Decimal

# 滑点分桶边界（bps）：0-5 / 5-15 / 15-30 / 30+
_SLIPPAGE_BUCKETS = (Decimal(5), Decimal(15), Decimal(30))
_BUCKET_LABELS = ("0-5", "5-15", "15-30", "30+")


@dataclass
class RunMetrics:
    runs: int = 0
    executed: int = 0
    rejected: int = 0
    not_approved: int = 0
    no_trade: int = 0
    errors: int = 0
    _slippage_sum_bps: Decimal = field(default=Decimal(0))
    _slippage_n: int = 0
    _slippage_hist: dict = field(default_factory=lambda: dict.fromkeys(_BUCKET_LABELS, 0))
    _approval_latency_sum: float = 0.0
    _approval_latency_max: float = 0.0
    _approval_latency_n: int = 0

    _STATUS_ATTR = {"executed": "executed", "rejected": "rejected",
                    "not_approved": "not_approved", "no_trade": "no_trade",
                    "execution_failed": "errors", "error": "errors"}

    def record(self, status: str, slippage_bps: Decimal | None = None) -> None:
        self.runs += 1
        attr = self._STATUS_ATTR.get(status, "errors")
        setattr(self, attr, getattr(self, attr) + 1)
        if slippage_bps is not None:
            self._slippage_sum_bps += slippage_bps
            self._slippage_n += 1
            for bound, label in zip(_SLIPPAGE_BUCKETS, _BUCKET_LABELS, strict=False):
                if slippage_bps < bound:
                    self._slippage_hist[label] += 1
                    break
            else:
                self._slippage_hist["30+"] += 1

    def record_approval_latency(self, seconds: float) -> None:
        self._approval_latency_sum += seconds
        self._approval_latency_max = max(self._approval_latency_max, seconds)
        self._approval_latency_n += 1

    @property
    def avg_slippage_bps(self) -> float:
        return float(self._slippage_sum_bps / self._slippage_n) if self._slippage_n else 0.0

    @property
    def approval_rate(self) -> float:
        decided = self.executed + self.not_approved
        return round(self.executed / decided, 3) if decided else 0.0

    @property
    def order_success_rate(self) -> float:
        attempted = self.executed + self.errors
        return round(self.executed / attempted, 3) if attempted else 0.0

    @property
    def avg_approval_latency_s(self) -> float:
        return round(self._approval_latency_sum / self._approval_latency_n, 3) \
            if self._approval_latency_n else 0.0

    def snapshot(self) -> dict:
        return {"runs": self.runs, "executed": self.executed, "rejected": self.rejected,
                "not_approved": self.not_approved, "no_trade": self.no_trade, "errors": self.errors,
                "avg_slippage_bps": round(self.avg_slippage_bps, 2), "approval_rate": self.approval_rate,
                "order_success_rate": self.order_success_rate,
                "slippage_hist_bps": dict(self._slippage_hist),
                "approval_latency": {"avg_s": self.avg_approval_latency_s,
                                     "max_s": round(self._approval_latency_max, 3),
                                     "n": self._approval_latency_n}}

File: internal/metrics/metrics.go
// Package metrics contains the in-process metrics snapshot consumed by the
// dashboard. Prometheus export can be layered on later without changing this
// compatibility contract.
package metrics

import (
	"math"
	"sync"
	"time"
)

// ApprovalLatency is kept numeric because the React dashboard calls toFixed.
type ApprovalLatency struct {
	AverageSeconds float64 `json:"avg_s"`
	MaxSeconds     float64 `json:"max_s"`
	Count          int     `json:"n"`
}

// RunSnapshot defines the stable /api/metrics payload.
type RunSnapshot struct {
	Runs               int             `json:"runs"`
	Executed           int             `json:"executed"`
	Rejected           int             `json:"rejected"`
	NotApproved        int             `json:"not_approved"`
	NoTrade            int             `json:"no_trade"`
	Errors             int             `json:"errors"`
	AverageSlippageBPS float64         `json:"avg_slippage_bps"`
	ApprovalRate       float64         `json:"approval_rate"`
	OrderSuccessRate   float64         `json:"order_success_rate"`
	SlippageHistogram  map[string]int  `json:"slippage_hist_bps"`
	ApprovalLatency    ApprovalLatency `json:"approval_latency"`
}

// Runs is safe for concurrent HTTP handlers and orchestrator goroutines.
type Runs struct {
	mu sync.RWMutex

	runs        int
	executed    int
	rejected    int
	notApproved int
	noTrade     int
	errors      int

	slippageSum float64
	slippageN   int
	histogram   map[string]int

	approvalSum time.Duration
	approvalMax time.Duration
	approvalN   int
}

func NewRuns() *Runs {
	return &Runs{histogram: map[string]int{"0-5": 0, "5-15": 0, "15-30": 0, "30+": 0}}
}

// Record records a terminal run status and optional execution slippage.
func (m *Runs) Record(status string, slippageBPS *float64) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.runs++
	switch status {
	case "executed":
		m.executed++
	case "rejected":
		m.rejected++
	case "not_approved":
		m.notApproved++
	case "no_trade":
		m.noTrade++
	default:
		m.errors++
	}
	if slippageBPS == nil || math.IsNaN(*slippageBPS) || math.IsInf(*slippageBPS, 0) {
		return
	}
	m.slippageSum += *slippageBPS
	m.slippageN++
	switch {
	case *slippageBPS < 5:
		m.histogram["0-5"]++
	case *slippageBPS < 15:
		m.histogram["5-15"]++
	case *slippageBPS < 30:
		m.histogram["15-30"]++
	default:
		m.histogram["30+"]++
	}
}

func (m *Runs) RecordApprovalLatency(duration time.Duration) {
	if duration < 0 {
		duration = 0
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	m.approvalSum += duration
	if duration > m.approvalMax {
		m.approvalMax = duration
	}
	m.approvalN++
}

func (m *Runs) Snapshot() RunSnapshot {
	m.mu.RLock()
	defer m.mu.RUnlock()

	avgSlippage := 0.0
	if m.slippageN > 0 {
		avgSlippage = m.slippageSum / float64(m.slippageN)
	}
	approvalRate := 0.0
	if decided := m.executed + m.notApproved; decided > 0 {
		approvalRate = float64(m.executed) / float64(decided)
	}
	orderSuccessRate := 0.0
	if attempted := m.executed + m.errors; attempted > 0 {
		orderSuccessRate = float64(m.executed) / float64(attempted)
	}
	avgApproval := 0.0
	if m.approvalN > 0 {
		avgApproval = m.approvalSum.Seconds() / float64(m.approvalN)
	}
	histogram := make(map[string]int, len(m.histogram))
	for key, value := range m.histogram {
		histogram[key] = value
	}

	return RunSnapshot{
		Runs: m.runs, Executed: m.executed, Rejected: m.rejected,
		NotApproved: m.notApproved, NoTrade: m.noTrade, Errors: m.errors,
		AverageSlippageBPS: round(avgSlippage, 2),
		ApprovalRate:       round(approvalRate, 3),
		OrderSuccessRate:   round(orderSuccessRate, 3),
		SlippageHistogram:  histogram,
		ApprovalLatency: ApprovalLatency{
			AverageSeconds: round(avgApproval, 3),
			MaxSeconds:     round(m.approvalMax.Seconds(), 3),
			Count:          m.approvalN,
		},
	}
}

func round(value float64, precision int) float64 {
	power := math.Pow10(precision)
	return math.Round(value*power) / power
}
